yk_gmd_blender/yk_gmd/structs/vertex_buffer_layout.py

```diff
@@ -22,50 +18,120 @@ class VertexBufferLayoutStruct(BaseBigEndianStructure):
+        if name == "unk":
+            result = self.get_vector_type(0x11)
+            return result if result != 4 else 3
+        if name == "col0":
+            return self.get_vector_type(0x16)
+        if name == "col1":
+            return self.get_vector_type(0x19)
+        if name == "uv":
+            if not vertex_packing & (1 << 0x1B):
+                # likely not to be UV
+                # treat as Vector2
+                return 1 << 16
+
+            # this assumes that uv_count can't be greater than 2
+            i = 0x1C
+            result = 0
+            uv_count = c_uint32(vertex_packing).value >> 0x1C
+            while uv_count > 0:
+                i += 4
+                shift = (vertex_packing >> i) & 0xF
+                if c_uint8(shift).value == 0xF:
+                    continue
+                uv_count -= 1
+                shift = (vertex_packing >> (i + 2)) & 3
+                if shift:
+                    if shift != 1:
+                        # add uv_count as index per type (highest comes first)
+                        # each 0x1 is a 4 * 8bit fixed point
+                        result += 1 + (uv_count << 4)
+                        continue
+                    # each 0x100 is a Vector2Half
+                    result += (1 << 8) + (uv_count << 12)
+                else:
+                    # each 0x10000 is a Vector2
+                    result += (1 << 16) + (uv_count << 20)
+            return result
 
     def get_vertex_layout(self) -> GMDVertexBufferLayout:
+        vertex_elems = {}
         vertex_layout_bits = [
-            ("pos", 0x03),
-            ("weights", 0x3c0),
-            ("normal", 0x1c00),
-            ("tangent", 0xe000),
-
-            ("col0", 0x00e0_0000),
-            ("col1", 0x0700_0000),
-            ("uv",   0x1800_0000),
+            ("pos", 0x07),
+            ("weights", 0x70),
+            ("bones", 0x200),
+            ("normal", 0x400),
+            ("tangent", 0x2000),
+            ("unk", 0x10000),
+            ("col0", 0x0020_0000),
+            ("col1", 0x0100_0000),
+            ("uv",   0xf000_0000),
         ]
-        overall_bitmask = 0
-        vertex_elems = {}
-        for (name, bitmask) in vertex_layout_bits:
-            overall_bitmask = overall_bitmask | bitmask
-            vertex_elems[name] = self.all_bits(int(self.vertex_packing), bitmask)
-        # overall_bitmask should equal 0x1FE0_FFC3
-        # there should also be no bits set outside of the bitmask
 
-        uv_bytes = 0
-        if self.uv_flags == 0x4:
-            uv_bytes = 2
-        elif self.uv_flags == 0x44:
-            uv_bytes = 4
+        for (name, bitmask) in vertex_layout_bits:
+            vertex_elems[name] = 0 if int(self.vertex_packing) & bitmask else -1
+        for name in vertex_elems:
+            if vertex_elems[name] == -1:
+                vertex_elems[name] = 0
+                continue
+            vertex_elems[name] = self.get_property_type(name)
```

Comment:
This is a little confusing - vertex_elems[name] is set to 0 or -1, but then if it was set to 0 before, it's set to something else, and if it was set to -1, it gets set to 0?

It would be easier to read if it were
```
for name, bitmask in vertex_layout_bits:
    if self.vertex_packing & bitmask:
        vertex_elems[name] = get_property_type(name)
    else:
        vertex_elems[name] = 0
```

Reply:
The functionality was different at first, but I forgot to change this. Applied your suggestion in the latest commit